pkg/serverless/interval.go

```diff
@@ -0,0 +1,68 @@
+package serverless
+
+import (
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/serverless/flush"
+)
+
+const maxInvocationsStored = 10
+
+// StoreInvocationTime stores the given invocation time in the list of previous
+// invocations. It is used to compute the invocation interval of the current function.
+// It is automatically removing entries when too much have been already stored (more than maxInvocationsStored).
+// When trying to store a new point, if it is older than the last one stored, it is ignored.
+// Returns if the point has been stored.
+func (d *Daemon) StoreInvocationTime(t time.Time) bool {
+	// ignore points less recent than the last stored one
+	if len(d.lastInvocations) > 0 && d.lastInvocations[len(d.lastInvocations)-1].After(t) {
+		return false
+	}
+
+	// remove when too much/old entries
+	d.lastInvocations = append(d.lastInvocations, t)
+	if len(d.lastInvocations) > maxInvocationsStored {
+		d.lastInvocations = d.lastInvocations[len(d.lastInvocations)-maxInvocationsStored : len(d.lastInvocations)]
+	}
+
+	return true
+}
+
+// InvocationInterval computes the invocation interval of the current function.
+// This function returns 0 if not enough invocations were done.
+func (d *Daemon) InvocationInterval() time.Duration {
+	// with less than 3 invocations, we don't have enough data to compute
+	// something reliable.
+	if len(d.lastInvocations) < 3 {
+		return 0
+	}
+
+	var total int64
+	for i := 1; i < len(d.lastInvocations); i++ {
+		total += int64(d.lastInvocations[i].Sub(d.lastInvocations[i-1]))
+	}
+
+	return time.Duration(total / int64(len(d.lastInvocations)-1))
+}
+
+// AutoSelectStrategy uses the invocation interval of the function to select the
+// best flush strategy.
+// This function doesn't mind if the flush strategy has been overridden through
+// configuration / environment var, the caller is responsible of that.
+func (d *Daemon) AutoSelectStrategy() flush.Strategy {
+	freq := d.InvocationInterval()
+
+	// when not enough data is available, fallback on flush.AtTheEnd strategy
+	if freq == time.Duration(0) {
+		return &flush.AtTheEnd{}
+	}
+
+	// if running more than 1 time every 5 minutes, we can switch to the flush strategy
+	// flushing at least every 10 seconds (at the start of the invocation)
+	// TODO(remy): compute a proper interval instead of hard-coding 10 seconds
+	if freq.Seconds() < 60*5 {
+		return flush.NewPeriodically(10 * time.Second)
+	}
```

Comment:
Could you explain why ? What is the gain of flushing 10s after the start of the function if I run every 3 minutes instead of 5 ?

Reply:
I think you've mixed up what we are trying to achieve.
Basically, the idea is to flush at the start of the function invocation instead of at the end, because it means that the flush will run in parallel with the function execution. But we can't do that for every function (hence the run interval detection) because there is scenario where we would never flush / discard some data if we would do so.

Here, we enable the "periodical" flush only with functions running often enough. We're not flushing 10s after the start of the function, we're flushing "at least every 10 seconds". Meaning, if the function is executed every 500ms for instance, the function would run 20 times in 10s, but we would do only one flush, at the start of the 20th invocation, avoiding to do too many useless flush (cpu time) and not losing any data. Let me know if that is still not clear, I can also share the RFC with you if needed.

Reply:
Ok, Thanks for the explanation.

Still why 5min ? Is it arbitrary ?

Reply:
It is based on timings measured by the Serverless team and with a few assumptions made together, so not exactly arbitrary even if a bit of a guess work for now.

Reply:
OK, thanks for the update. Do we want to add a comment explaining the reasoning ?

Reply:
Yeah good idea, I'm on it.